### PR TITLE
Rewrite minimum distance algorithm

### PR DESCRIPTION
**Pull request by dschwen**
Shorter minimum bubble distance algorithm. Uses fewer local variables (I eliminated 3!) and does not discard a viable center location if it happens to be found during the last try!

To break out of nested loops I'm using goto here. The jump distances are extremely short and documented using comments. I think the gains outweigh the potential reservations some people may have against the goto keyword.

Refs #7296

### CODE REVIEW

**Comment by FHilty** [comment]
I like the changes, it should perform the same job as it did before so far as i can tell. One thing though n the `initialSetup` block there is a switch with only one case, would something like this work better?

```
      if (_radius_variation > 0.0 && radius_variation_type == 2)
        mooseError("If radius_variation > 0.0, you must pass in a radius_variation_type in MultiSmoothCircleIC");
```

**Comment by dschwen** [comment]
Yeah, that should probably get changed as well.

**Comment by dschwen** [comment]
Also refactored the super ellipsoid collision now.

**Comment by dschwen** [comment]
I'm adding the bimodal stuff as well

File: modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C [hunk 1]
```diff
@@ -124,9 +123,8 @@ BimodalSuperellipsoidsIC::computeSuperellipsoidExponents()
     _ns[i] = _small_n;
 }
 
-
 void
-BimodalSuperellipsoidsIC::computeSuperellipsoidCenters()
+BimodalSuperellipsoidsIC::computeCircleCenters()
```
Comment (laagesen) [inline modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C]: why the change to computeCircleCenters?

Comment (dschwen) [inline modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C]: It is a wip commit. I'm still working on this.

Comment (laagesen) [inline modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C]: Ah Ok.

File: modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C [hunk 2]
```diff
@@ -58,20 +59,18 @@ MultiSmoothSuperellipsoidIC::MultiSmoothSuperellipsoidIC(const InputParameters &
 void
```
Comment (FHilty) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: In the class parameters above here there is still a `bool` for `_prevent_overlap`, how the code is currently set up it has no function as the `ellipsoidsOverlap` function is always utilized. I had left it as an option because if a large enough `bubspac` is used this check becomes redundant. However for my purposes I will always use `ellipsoidsOverlap` anyway so this change won't affect my work.

**Comment by laagesen** [comment]
`MultiSmoothCircle` and `MultiSmoothSuperellipsoid` look good.

**Comment by dschwen** [comment]
I'll create a common base class that will do the center randomization and collision checking. There is too much duplicated code currently. This will need a small Marmot fix.

**Comment by permcody** [comment]
BASE CLASS

**Comment by dschwen** [comment]
No base class for now. There are some subtle differences that would make this a pain

**Comment by dschwen** [comment]
Marmot patch is ready!

File: modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C [hunk 3]
```diff
@@ -67,34 +66,34 @@ BimodalSuperellipsoidsIC::computeSuperellipsoidSemiaxes()
...
-  for (unsigned int i = _input_as.size(); i < _input_as.size() + _npart; i++)
+  // Then fill in the randomly positioned (small) superellipsoids
+  for (unsigned int i = _input_as.size(); i < _input_as.size() + _npart; ++i)
   {
-    //Vary semiaxes
+    // Vary semiaxes
     switch (_size_variation_type)
     {
-      case 0: //Random distrubtion, maintaining constant shape
+      case 0: //R andom distrubtion, maintaining constant shape
```
Comment (permcody) [inline modules/phase_field/src/ics/BimodalSuperellipsoidsIC.C]: This is a somewhat "random" change.

File: modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C [hunk 4]
```diff
@@ -58,20 +59,18 @@ MultiSmoothSuperellipsoidIC::MultiSmoothSuperellipsoidIC(const InputParameters &
...
 MultiSmoothSuperellipsoidIC::initialSetup()
 {
-  Real nv = _numbub.size();
+  unsigned int nv = _numbub.size();
```
Comment (permcody) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: May as well go "auto" 😄

Comment (dschwen) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: Not without the crazy _auto/decltype_ syntax in the for loop below.
